Fix droplist completer focus

vidartf opened the pull request:
Currently, if you click on the `CompletionWidget` with the mouse, it gets focus (Windows 10, 64-bit, PyQt4). This triggers the `CompletionWidget` to hide. To prevent this, check the active window before hiding. Next, since it now has focus, any keyboard input to the `CompletionWidget` needs to be redirected to its `_text_editor`, except the input that is intended for it (up/down/page up/page down/home/end).

vidartf commented:
This should also fix spyder-ide/spyder#1908 and spyder-ide/spyder#2265.

ccordoba12 commented:
I've found a problem with your pull request. Steps to reproduce:

1. Write `import <Tab>` and select an entry in the completion widget with a **single** mouse click.
2. Start to write some text right after doing a selection, e.g. `set`
3. Do a **single** mouse click out of qtconsole (e.g. in Chrome or Firefox).
4. You can see the completion widget is not hidden (as it should be) but instead continues to be above qtconsole.

vidartf commented:
Fixed. Rebased, so anything based off the last one is detached. Now does some magic ritual in the initialization that I'm not 100% sure of what does, but I can hazard a guess. Now it also sets the width of the widget correctly (and therefore forces horz scrollbar off).

vidartf commented:
This now works smoothly on Windows, but it would be very helpful if someone on other systems could verify.

takluyver commented:
Just tested on Linux (anyone else testing, the option you need is `--gui-completion=droplist`). Unfortunately, with this PR, the dropdown for completions doesn't appear at all. It's still making the completion request, and if there's only one possible completion, it fills it in, but if there's more than one, no dice.

No error messages when this happens.

vidartf commented:
Would you mind playing with `git reset` to see at which commit this start happening?

takluyver commented:
I used git bisect, and it say 8917eeb0683e9047f7dc55051c65cdae0f6ee7ed is where it went wrong:

```
8917eeb0683e9047f7dc55051c65cdae0f6ee7ed is the first bad commit
commit 8917eeb0683e9047f7dc55051c65cdae0f6ee7ed
Author: Vidar Tonaas Fauske <vidartf@gmail.com>
Date:   Sun Jan 24 04:30:56 2016 +0100

    Add some CompletionWidget setup magic
```

vidartf commented:
PyQt or PySide?

takluyver commented:
Looks like it's picking up PyQt5 on my system.

vidartf commented:
Could you try to run the tests I checked in and see which i any succeed?

takluyver commented:
They all pass (I see another failure, but I see the same in master, and I think it's unrelated).

vidartf commented:
Could you try re-adding the  `QtCore.Qt.WindowStaysOnTopHint` window flag?

takluyver commented:
I still don't see the dropdown.

vidartf commented:
Well, as far as I can understand, it should be the code in `__init__` that caused the change. Of that code it could either be the window flags (popup vs tooltip), focus policy, or the four lines that were added to the start (or a combination of these). If you'd be able to revert thing piece by piece, we could maybe figure out why it doesn't work.

takluyver commented:
It's the window flags - if I change `Popup` back to `ToolTip`, it works.

vidartf commented:
Could you try adding the following statement somewhere in the init?
```python
self.setAttribute(QtCore.Qt.WA_X11NetWmWindowTypeCombo)
```

vidartf commented:
... and/or adding this to the end of `show_items()` just before `self.show()`: `self.raise_()` (since all the tests seems to indicate that it is shown, it might just have the wrong Z-order?).

vidartf commented:
If neither of those work, we could try to add a platform conditional for which flag(s) to use? Is there a standard way for this project for how to check the platform?

ccordoba12 commented:
> Is there a standard way for this project for how to check the platform?

Python has it:

1. `os.name == 'nt'`: Windows
2. `sys.platform == 'darwin'`: OS X
3. `sys.platform.startswith('linux')`: Linux

takluyver commented:
Neither of those did the trick, either singly or in combination.

Thanks for your persistence with this!

vidartf commented:
Ok, added platform specific flags. The PR should still fix the issue of horizontal scroller / popup width as discussed in spyder-ide/spyder#2265. Care to confirm that everything works as advertised?

Note: I assumed the previous behavior worked on any OS other than Windows. If that is not the case, it should be easy to tune.

takluyver commented:
That seems to be working for me. Thanks!

takluyver commented on qtconsole/completion_widget.py:
```diff
@@ -20,9 +21,21 @@ def __init__(self, console_widget):
+        self.setSelectionBehavior(QtGui.QAbstractItemView.SelectRows)
+        self.setSelectionMode(QtGui.QAbstractItemView.SingleSelection)
+
+        if os.name == 'nt':
```
Let's leave a brief comment here explaining why we need different flags for different systems.

We should also try to get someone with OSX to check that it's working for them.

vidartf replied:
Pushed comments. If there's someone on OSX that can test this in time for the appropriate milestone, then all the better, if not, it does at least fix an existing issue for Windows users.

ccordoba12 commented on qtconsole/completion_widget.py:
```diff
@@ -20,9 +21,28 @@ def __init__(self, console_widget):
 
         self.setAttribute(QtCore.Qt.WA_StaticContents)
-        self.setWindowFlags(QtCore.Qt.ToolTip | QtCore.Qt.WindowStaysOnTopHint)
+        origPolicy = text_edit.focusPolicy()
```
Please change this to `original_policy` to follow the Python custom of naming variables with underscores instead of camelcase.

ccordoba12 commented on qtconsole/tests/test_completion_widget.py:
```diff
@@ -0,0 +1,70 @@
+elif api == QT_API_PYQT5:
+    from PyQt5.QtTest import QTest
+elif api == QT_API_PYSIDE:
+    from PySide.QtTest import QTest
```
I think this could be simplified by only importing `QT_API_PYQT5` and `QT_API_PYSIDE` and writing this block as

```python
if api == QT_API_PYQT5:
    from PyQt5.QtTest import QTest
elif api == QT_API_PYSIDE:
    from PySide.QtTest import QTest
else:
    from PyQt4.QtTest import QTest
```

ccordoba12 commented:
@vidartf, I'm sorry to say it but your changes doesn't work for me on Linux.

If I select an entry in the completion widget with a mouse right-click, it hides immediately, just as before.

ccordoba12 commented:
Things were working when I tested your branch last week though :-s

ccordoba12 commented:
I should have tested further: the failure I reported only appears on PyQt5. On PyQt4 your changes work as expected.

vidartf commented:
@ccordoba12, on PyQt5, could you test with using the `Qt.Popup` flags? Maybe only linux + PyQt4 needs `Qt.ToolTip`?

ccordoba12 commented:
@vidartf, that's not the problem. I used `Qt.Popup` instead of `Qt.ToolTip`and the widget doesn't show up (as reported by @takluyver).

vidartf commented:
Does linux + PyQt5 work with the old code?

ccordoba12 commented:
Not, it doesn't. The completion widget in not displayed with your changes from 8 days ago.

vidartf commented:
Not sure if we talked past each other or not. I meant: Does droplist work with linux + PyQt5 on master?

ccordoba12 commented:
> Does droplist work with linux + PyQt5 on master?

Yes, the completion widget is shown as expected.

vidartf commented:
> Yes, the completion widget is shown as expected.

And can be clicked?

ccordoba12 commented:
Nop, when you click on any entry the widget disappears immediately (but 
that is also the case with PyQt4 on master).

El 01/02/16 a las 09:50, Vidar Tonaas Fauske escribió:
>
>     Yes, the completion widget is shown as expected.
>
> And can be clicked?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/jupyter/qtconsole/pull/93#issuecomment-178000870>.
>

vidartf commented:
Oh. From spyder-ide/spyder#1908 I got the impression this was a Windows specific issue. Umm.. In that case I'm kind of out of ideas, except that you switch to `Popup` and try to figure out what is needed to get it to show on Linux? Maybe set the parent? Currently the completer doesn't have a parent (from the git log, somebody complained about segfaults on closing), so you might try setting that?

If you try the bottom-up approach, maybe test that you can get a generic QWidget/QWindow show with the `Popup` flag set, then switch to a `QListItemView`, and then progressively get closer to the flags/attributes of the completion widget. Or you could do top-down and see what needs to be removed for it to show.

Either way, it's kind of hard for me to debug as I don't currently have access to a linux machine.

ccordoba12 commented:
Could you point to the commit that removed parent?

vidartf commented:
https://github.com/ipython/ipython/commit/f21a128a37a6f32ed502e2fe5bf4e82a35875043

ccordoba12 commented:
@vidartf, I sent a pull request against your branch that fixes the issues I'm seeing with PyQt5.

Please test it on Windows to see if it didn't break anything there :-)

vidartf commented:
I've merged the proposed changed from @ccordoba12. Would you mind confirming whether this still works on your setup @takluyver ?

takluyver commented:
That appears to be working for me with PyQt5.

vidartf commented:
Alright, I cleaned up some more code, and think this can be reviewed for merge again. If possible, it should be tested with PySide and on Mac OSX.

takluyver commented:
Checked under PySide + Python 2 on Linux (I was testing with Python 3 before).

@Carreau have you got a few minutes to check this PR on a Mac and check that it hasn't broken anything? You'll need to run with `--gui-completion=droplist` to test it.

Carreau commented:
> have you got a few minutes to check this PR on a Mac and check that it hasn't broken anything? You'll need to run with --gui-completion=droplist to test it.

Seem to work for me.

ccordoba12 commented:
Ok, everything seems to be working fine for me too, on both PyQt4 and PyQt5 with Linux.

Thanks a lot for your work on this @vidartf!
